### PR TITLE
feat(net/peer): add peer with udp socket

### PR DESCRIPTION
**Pull request by jsvisa**
ref https://github.com/paradigmxyz/reth/pull/9043#pullrequestreview-2144882355

@mattsse could you please a look at this one first

### CODE REVIEW

File: crates/net/network/src/discovery.rs [hunk 1]
```diff
@@ -40,7 +40,7 @@ pub struct Discovery {
     /// All nodes discovered via discovery protocol.
     ///
     /// These nodes can be ephemeral and are updated via the discovery protocol.
-    discovered_nodes: LruMap<PeerId, SocketAddr>,
+    discovered_nodes: LruMap<PeerId, (SocketAddr, SocketAddr)>,
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:43]: this should document the addresses.

I think we should consider a helper type in the peers crate like `PeerAddr {tcp, udp}`

File: crates/net/peers/src/node_record.rs [hunk 2]
```diff
@@ -92,6 +92,17 @@ impl NodeRecord {
         Self { address: addr.ip(), tcp_port: addr.port(), udp_port: addr.port(), id }
     }
 
+    /// Creates a new record from an ip address and ports.
+    pub fn new_with_ip_ports(
```
Comment (mattsse) [inline crates/net/peers/src/node_record.rs:96]:
```suggestion
    pub fn new_with_ports(
```

File: crates/net/network/src/peers.rs [hunk 3]
```diff
@@ -325,33 +355,24 @@ impl PeersManager {
...
                 is_trusted = is_trusted || peer.is_trusted();
-
-                // if a peer is not trusted and we don't have capacity for more inbound connections,
-                // disconnecting the peer
-                if !is_trusted && !has_in_capacity {
-                    self.queued_actions.push_back(PeerAction::Disconnect {
-                        peer_id,
-                        reason: Some(DisconnectReason::TooManyPeers),
-                    });
-                }
```
Comment (mattsse) [inline crates/net/network/src/peers.rs:336]: what happened here?

File: crates/net/network/src/peers.rs [hunk 4]
```diff
@@ -997,8 +1031,10 @@ impl ConnectionInfo {
 /// Tracks info about a single peer.
 #[derive(Debug, Clone)]
 pub struct Peer {
-    /// Where to reach the peer
-    addr: SocketAddr,
+    /// Peer's TCP address.
+    tcp_addr: SocketAddr,
+    /// Peer's UDP address if it's port is different to the TCP's.
```
Comment (mattsse) [inline crates/net/network/src/peers.rs:1036]: hmm we don't store this conditionally, so the comment is wrong?

File: crates/net/network-api/src/lib.rs [hunk 5]
```diff
@@ -67,8 +67,8 @@ pub trait PeersInfo: Send + Sync {
 /// Provides an API for managing the peers of the network.
 pub trait Peers: PeersInfo {
     /// Adds a peer to the peer set.
-    fn add_peer(&self, peer: PeerId, addr: SocketAddr) {
-        self.add_peer_kind(peer, PeerKind::Basic, addr);
+    fn add_peer(&self, peer: PeerId, tcp_addr: SocketAddr, udp_addr: Option<SocketAddr>) {
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:70]: these changes are a bit invasive, these changes can be prevented by adding another set of functions that also accept the `udp: SocketAddr` which then also delegate to add_peer_kind

**Review by mattsse — CHANGES_REQUESTED** [review]
supportive

File: crates/net/network/src/peers.rs (hunk 3 above)
Comment (jsvisa) [reply crates/net/network/src/peers.rs:336]: sorry, this logic is duplicated, I should modify them in another PR, not in this one.

File: crates/net/network-api/src/lib.rs (hunk 5 above)
Comment (jsvisa) [reply crates/net/network-api/src/lib.rs:70]: thanks for the tips, this really helpful, no need to adjust many test files.

File: crates/net/network/src/peers.rs (hunk 4 above)
Comment (jsvisa) [reply crates/net/network/src/peers.rs:1036]: the udp is option, i think you're looking at the upper line.

**Comment by jsvisa** [comment]
@mattsse I have one more concern, do you think which one is good:

1. `add_peer_xxx` API in  Peer Trait receives two `SocketAddr`, and inside net/network use `PeerAddr` to represent the inner state;
2. `add_peer_xxx` receives `PeerAddr` instead

I'd prefer to the first one, so `PeerAddr` can represent and used inner

**Comment by mattsse** [comment]
> I'd prefer to the first one, so PeerAddr can represent and used inner

same

**Comment by jsvisa** [comment]
@mattsse i think this pr is ready to review, please take a look.

File: crates/net/network-api/src/lib.rs [hunk 6]
```diff
@@ -77,12 +82,18 @@ pub trait Peers: PeersInfo {
     fn add_trusted_peer_id(&self, peer: PeerId);
 
     /// Adds a trusted peer to the peer set.
-    fn add_trusted_peer(&self, peer: PeerId, addr: SocketAddr) {
-        self.add_peer_kind(peer, PeerKind::Trusted, addr);
+    fn add_trusted_peer(&self, peer: PeerId, tcp_addr: SocketAddr, udp_addr: Option<SocketAddr>) {
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:85]: this should also be two functions, as as the regular add_peer

File: crates/net/network/src/peers.rs [hunk 7]
```diff
@@ -338,7 +362,7 @@ impl PeersManager {
             Entry::Vacant(entry) => {
                 // peer is missing in the table, we add it but mark it as to be removed after
                 // disconnect, because we only know the outgoing port
-                let mut peer = Peer::with_state(addr, PeerConnectionState::In);
+                let mut peer = Peer::with_state(PeerAddr::new(addr, None), PeerConnectionState::In);
```
Comment (mattsse) [inline crates/net/network/src/peers.rs:365]: let's add a `PeerAddr::tcp(addr)` constructor fn as well

File: crates/net/network/src/peers.rs
Comment (mattsse) [inline crates/net/network/src/peers.rs:1]: please use `PeerAddr::tcp` over `new(addr, None)`

**Review by mattsse — CHANGES_REQUESTED** [review]
a few more style nits.

in general, `Option` in `new` functions can/should always be avoided if the type if the type can be created without having the optional input.

**Comment by jsvisa** [comment]
> in general, Option in new functions can/should always be avoided if the type if the type can be created without having the optional input.

Hi @mattsse, a small curious about is a Rust style or Reth's style, as in other languages, eg Python with default argument values, so can use None as the default value. 

As we also need the Option udp value to instantiation the PeerAddr, eg https://github.com/paradigmxyz/reth/blob/5b2a05b7ac9d9c1934de1db6e2238a6c93bdbcd4/crates/net/network/src/network.rs#L266-L273, so I keep the `PeerAddr::new(TCP, Option<UDP>)` func sign.

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
just my personal preference, I find it easier to read if there's a constructor that does not accept Option, because usually you can also combine them with builder style functions like `new(a).with(b)..`

**Pull request merged**